File: cses/04_graph_algorithms/1669_round-trip/solution.py
import sys


def main():
    # Đọc toàn bộ input một lần cho nhanh.
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    pos = 0
    n = int(data[pos]); pos += 1
    m = int(data[pos]); pos += 1

    # Dựng danh sách kề cho đồ thị VÔ HƯỚNG (mỗi cạnh thêm hai chiều).
    adj = [[] for _ in range(n + 1)]
    for _ in range(m):
        a = int(data[pos]); b = int(data[pos + 1]); pos += 2
        adj[a].append(b)
        adj[b].append(a)

    visited = bytearray(n + 1)
    parent = [0] * (n + 1)   # 0 nghĩa là không có parent (đỉnh đánh số 1..n)
    ptr = [0] * (n + 1)      # con trỏ duyệt danh sách kề cho DFS lặp

    result = None

    # Duyệt từng thành phần liên thông bằng DFS lặp cho tới khi tìm được chu trình.
    for s in range(1, n + 1):
        if visited[s]:
            continue
        visited[s] = 1
        parent[s] = 0
        stack = [s]
        while stack:
            u = stack[-1]
            adj_u = adj[u]
            if ptr[u] < len(adj_u):
                v = adj_u[ptr[u]]
                ptr[u] += 1
                if v == parent[u]:
                    # Bỏ qua đúng một cạnh quay lại parent (không có cạnh bội).
                    continue
                if not visited[v]:
                    # Cạnh cây: đi sâu xuống v.
                    visited[v] = 1
                    parent[v] = u
                    stack.append(v)
                else:
                    # Back edge u -> v: v là tổ tiên của u. Truy ngược theo
                    # parent từ u về v để dựng chu trình [u, ..., v, u].
                    cyc = [u]
                    x = u
                    while x != v:
                        x = parent[x]
                        cyc.append(x)
                    cyc.append(u)
                    result = cyc
                    break
            else:
                # Đã duyệt hết cạnh kề của u -> lấy u ra khỏi stack.
                stack.pop()
        if result is not None:
            break

    # Không tìm thấy chu trình nào -> in "IMPOSSIBLE".
    out = sys.stdout
    if result is None:
        out.write("IMPOSSIBLE\n")
    else:
        out.write(str(len(result)) + "\n")
        out.write(" ".join(map(str, result)) + "\n")


main()

File: cses/04_graph_algorithms/1669_round-trip/solution.cpp
#include <cstddef>
#include <iostream>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n, m;
    std::cin >> n >> m;

    // Dựng danh sách kề cho đồ thị VÔ HƯỚNG (mỗi cạnh thêm hai chiều).
    std::vector<std::vector<int>> adj(n + 1);
    for (int i = 0; i < m; ++i) {
        int a, b;
        std::cin >> a >> b;
        adj[a].push_back(b);
        adj[b].push_back(a);
    }

    // state: 0 = chưa thăm, 1 = đang nằm trên stack DFS, 2 = đã duyệt xong.
    std::vector<int> state(n + 1, 0);
    // parent[v] = đỉnh mà từ đó ta đi đến v trên cây DFS.
    std::vector<int> parent(n + 1, 0);
    // next_edge[u] = con trỏ duyệt danh sách kề của u cho DFS lặp.
    std::vector<std::size_t> next_edge(n + 1, 0U);
    std::vector<int> cycle;

    // Duyệt từng thành phần liên thông cho tới khi tìm được một chu trình.
    for (int start = 1; start <= n && cycle.empty(); ++start) {
        if (state[start] != 0) {
            continue;
        }

        std::vector<int> stack{start};
        state[start] = 1;
        parent[start] = 0;

        while (!stack.empty() && cycle.empty()) {
            const int u = stack.back();
            // Đã duyệt hết cạnh kề của u -> hoàn tất, lấy u ra khỏi stack.
            if (next_edge[u] == adj[u].size()) {
                state[u] = 2;
                stack.pop_back();
                continue;
            }

            const int v = adj[u][next_edge[u]++];
            // Bỏ qua đúng cạnh quay lại parent (an toàn vì không có cạnh bội).
            if (v == parent[u]) {
                continue;
            }
            if (state[v] == 0) {
                // Cạnh cây: đi sâu xuống v.
                state[v] = 1;
                parent[v] = u;
                stack.push_back(v);
            } else if (state[v] == 1) {
                // Back edge u -> v: v là tổ tiên còn trên stack. Truy ngược
                // theo parent từ u về v để dựng chu trình [u, ..., v, u].
                cycle.push_back(u);
                int current = u;
                while (current != v) {
                    current = parent[current];
                    cycle.push_back(current);
                }
                cycle.push_back(u);
            }
        }
    }

    // Không gặp back edge nào -> đồ thị không có chu trình.
    if (cycle.empty()) {
        std::cout << "IMPOSSIBLE\n";
        return 0;
    }

    // In số đỉnh của chu trình rồi tới dãy đỉnh.
    std::cout << cycle.size() << '\n';
    for (std::size_t i = 0; i < cycle.size(); ++i) {
        if (i != 0U) {
            std::cout << ' ';
        }
        std::cout << cycle[i];
    }
    std::cout << '\n';
    return 0;
}
